### PR TITLE
Modify mesonintrospect --dependencies so 1 element in list is 1 depen…

### LINKED ISSUE #2018: mesonintrospect --dependencies returns list instead of dictionary
Reported it breaking completely in #1963 .

When I got around to testing the fix just now I noticed that the root object for mesonintrospect --dependencies now is a list instead of an object like it used to be prior to 0.41. Probably not what is wanted.

Example output (added newlines to make it easier to read):

```shell
$ mesonintrospect --dependencies <build path>
["threads", {"compile_args": [], "link_args": []},
 "libjpeg", {"compile_args": [], "link_args": ["-ljpeg"]},
 "libpng", {"compile_args": ["-I/usr/include/libpng16"], "link_args": ["-lpng16"]},
 "libwebp", {"compile_args": [], "link_args": ["-lwebp"]}]
```

### PR DESCRIPTION
**Pull request by martin-ejdestig**
…dency

Makes it less awkward for tools to parse the data.

Was modified to return a list, with two elements for each dependency,
instead of a dictionary in ac1c929f66bde7209f1bd7e2d995dbd3949e1d8b .

Closes #2018.

### CODE REVIEW

**Comment by jpakkane** [comment]
If there are two dependencies with the same name but different modules (such as Qt5+Core and Qt5+Qml) then this will produce two items with the same name but different args. That is confusing. Could we make the output clearer in this case?

**Comment by martin-ejdestig** [comment]
Add a "modules" key? Add module names to dependency name?

@nirbheek Any ideas?

(It is also not clear to me why the concept of modules are needed. Why are they not separate dependencies? I.e. they have separate pkg-config files. Convenience?)

**Comment by martin-ejdestig** [comment]
It also feels like a bit of a separate issue. The main intention with the commit was to get rid of the inconvenient part where two elements in the list are used to describe one dependency.

@jpakkane Or maybe you mean that we can go back to the pre 0.41 format where the root object is a dict instead if we solve the same name but different modules clash part?

**Comment by nirbheek** [comment]
Yes, I agree. We can add arbitrary data to the list later, such as the list of modules. However, it would be good to add some unit tests for this? To ensure that the feature is working at all.

**Comment by martin-ejdestig** [comment]
Tests would be good, yes. How/where should I add mesonintrospect tests? There are none at the moment?

**Comment by nirbheek** [comment]
See `run_unittests.py`. It has `mesonintrospect` tests.

**Comment by martin-ejdestig** [comment]
What external dependency/dependencies can be relied upon in the tests? Or should they be mocked somehow?

**Comment by jpakkane** [comment]
Look into `test cases/frameworks`.

**Comment by nirbheek** [comment]
Most of the framework dependencies can be expected to exist on Linux. You can assume that any dependency tests that are not optional will always be available.

**Comment by martin-ejdestig** [comment]
I added a test in LinuxlikeTests that checks for glib and gobject.

File: run_unittests.py [hunk 1]
```diff
@@ -1815,6 +1815,32 @@ def test_order_of_l_arguments(self):
...
         raise RuntimeError('Linker entries not found in the Ninja file.')
 
+    def test_introspect_dependencies(self):
+        '''
+        Tests that mesonintrospect --dependencies returns expected output.
+        '''
+        if not shutil.which('pkg-config'):
+            raise unittest.SkipTest('pkg-config not found')
+        if subprocess.call(['pkg-config', '--exists', 'glib-2.0', 'gobject-2.0']) != 0:
+            raise unittest.SkipTest('GLib and GObject not found with pkg-config')
```
Comment (nirbheek) [inline run_unittests.py]: You don't need these checks. We require these to exist on Linux for the project tests.

Comment (martin-ejdestig) [inline run_unittests.py]: I looked at what LinuxlikeTests.test_qt5dependency_pkgconfig_detection() does. Qt is not a requirement but GTK+ is?

Comment (nirbheek) [inline run_unittests.py]: Neither GTK+ nor Qt are requirements, but glib is.

Comment (martin-ejdestig) [inline run_unittests.py]: Or can the checks be removed from test_qt5dependency_pkgconfig_detection() as well?

Comment (martin-ejdestig) [inline run_unittests.py]: Ah, check.

Comment (martin-ejdestig) [inline run_unittests.py]: Done.
